spec/helpers.lua

```diff
@@ -1390,66 +1390,340 @@ local function wait_until(f, timeout, step)
-  local body = luassert.res_status(200, res)
-  local json = cjson.decode(body)
+---get routes | services | consumers | upstreams | targets | plugins
+---@param entity string routes | services | consumers | upstreams | targets | plugins
+---@param request_path? string path of Admin API
+---@param upstream_id? string optional, set only if `entity` is `targets`
+---@param admin_client_timeout? any
+---@param forced_admin_port? any
+---@return table entities
+---@return string next_request_path
+local function get_config(entity, request_path, upstream_id, admin_client_timeout, forced_admin_port)
+  if  entity ~= "routes"
+  and entity ~= "services"
+  and entity ~= "consumers"
+  and entity ~= "upstreams"
+  and entity ~= "targets"
+  and entity ~= "plugins"
+  and type(request_path) ~= "string"
+  then
+    error("arg #1 must be one of `routes`, `services`, `consumers`, `upstreams`, and `targets`")
+  end
 
-  return assert(json.timers)
-end
+  if entity == "targets" and type(upstream_id) ~= "string" then
+    error("arg #3 must be a string if arg#1 is `targets")
+  end
 
+  if not request_path then
+    if entity == "targets" then
+      request_path = "/upstreams/" .. upstream_id .. "/targets"
+
+    else
+      request_path = "/" .. entity
+    end
+  end
 
----wait for all timers created after `wait_timers_begin` was called to finish
----@param ctx table was returned by `wait_timers_begin`
----@param timeout number optional, maximum time to wait (default = 2)
----@param admin_client_timeout? number optional, to override the default timeout setting
----@param forced_admin_port? number optional, to override the default port of admin API
-local function wait_timers_end(ctx, timeout,
-                                   admin_client_timeout, forced_admin_port)
   local _admin_client = admin_client(admin_client_timeout, forced_admin_port)
-  local res = assert(_admin_client:get("/timers"))
+  local res = assert(_admin_client:get(request_path))
   local body = luassert.res_status(200, res)
-  local json = cjson.decode(body)
+  local json = assert(cjson.decode(body))
+
+  _admin_client:close()
+
+  if not json.data then
+    return nil, nil
+  end
+
+  if type(json.next) == "string" then
+    return json.data, json.next
+  end
+
+  return json.data, nil
+end
+
+
+---compare two values. if they are tables, then compare their keys and fields recursively.
+---@param t1 any
+---@param t2 any
+---@return boolean is_equal
+---@usage `deepcompare({ a = 10, b = 20 }, { a = 10, b = 20 })` -> `true`
+---@usage `deepcompare({ a = 10, b = 20 }, { a = 10, b = 30 })` -> `false`
+---@usage `deepcompare({ a = 10, b = 20 }, { a = 10, b = "" })` -> `false`
+---@usage `deepcompare({ a = 10, b = 20 }, { a = 10 })` -> `false`
+---@usage `deepcompare({ a = 10, b = { 1, 2, 3 } }, { a = 10, b = { 3, 2, 1 } })` -> `true`
+---@usage `deepcompare({ a = 10, b = { 1, 2, 3 } }, { a = 10, b = { 3, 2, 1, 0 } })` -> `false`
+local function deepcompare(t1, t2)
```

Comment:
can we use pl.tablex (deepcompare and merge) for this and following function, to avoid the need to test this function as well?